File: 3.regresja/regression.py
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
import pandas as pd

df = pd.read_csv("dane_regresja.csv", sep=';', decimal=',')

# kopiuje i wycinam wczorajszą datę z danych, aby zrobić na niej prognozę
df_today = df[df['Data'] == '2023-10-24'].copy()
df.drop(index=df_today.index, inplace=True)

df_today = df_today.reset_index(drop=True)

# dane objaśniające: zapo, wiat, PV
X = df[['Zapotrzebowanie', 'Wiatr', 'PV', 'Rezerwa']]
# dane onjaśniane: cena
y = df['RCE']

reg = LinearRegression().fit(X, y)
# R kwadrat
print("R kwadrat:", reg.score(X, y))

# predykcja cen na jutro
today_predict = reg.predict(df_today[['Zapotrzebowanie', 'Wiatr', 'PV', 'Rezerwa']])
today_predict = pd.Series(today_predict, name='RCE_pred')
today_predict = pd.merge(df_today['RCE'], today_predict, left_index=True, right_index=True)
print(today_predict)
sns.lineplot(today_predict[['RCE', 'RCE_pred']])
plt.title("Prognoza regresji liniowej vs wykonanie z dnia 2023-10-24")
plt.show()

File: 3.regresja/regression.R
# Instalacja i załadowanie potrzebnych pakietów
library(tidyverse)
library(caret)

# Wczytanie danych
df <- read.csv("C:\\UKSW\\Analityka-danych-w-badaniach-spolecznych\\3.regresja\\dane_regresja.csv", sep=';', dec=',')

# Skopiowanie i usunięcie wczorajszej daty z danych dla prognozy
df_today <- df[df$Data == '2023-10-24', ]
df <- df[df$Data != '2023-10-24', ]

# Dane objaśniające: Zapotrzebowanie, Wiatr, PV, Rezerwa
X <- df[c('Zapotrzebowanie', 'Wiatr', 'PV', 'Rezerwa')]
# Dane objaśniane: Cena
y <- df$'RCE'

# Dopasowanie modelu regresji liniowej
reg <- lm(y ~ ., data = cbind(y, X))

# R kwadrat
cat("R kwadrat:", summary(reg)$adj.r.squared, "\n")

# Predykcja cen na jutro
today_predict <- predict(reg, newdata = df_today[c('Zapotrzebowanie', 'Wiatr', 'PV', 'Rezerwa')])
today_predict <- data.frame(RCE_actual = df_today$`RCE`, RCE_pred = today_predict)

# Wykres prognozy regresji liniowej a wykonania z dnia 2023-10-24
with(today_predict, plot(RCE_actual, type = "l", col = "blue", lwd = 2, ylim = range(c(RCE_actual, RCE_pred)), main = "Prognoza regresji liniowej vs wykonanie z dnia 2023-10-24"))
lines(today_predict$RCE_pred, col = "red", lwd = 2)
legend("topright", legend = c("RCE_actual", "RCE_pred"), col = c("blue", "red"), lty = 1, lwd = 2)